hack/make-rules/test.sh

```diff
@@ -180,52 +172,45 @@ junitFilenamePrefix() {
   echo "${KUBE_JUNIT_REPORT_DIR}/junit_$(kube::util::sortable_date)"
 }
 
-produceJUnitXMLReport() {
-  local -r junit_filename_prefix=$1
-  if [[ -z "${junit_filename_prefix}" ]]; then
-    return
-  fi
-
-  local junit_xml_filename
-  junit_xml_filename="${junit_filename_prefix}.xml"
-
+installTools() {
   if ! command -v gotestsum >/dev/null 2>&1; then
     kube::log::status "gotestsum not found; installing from ./hack/tools"
     go -C "${KUBE_ROOT}/hack/tools" install gotest.tools/gotestsum
   fi
-  gotestsum --junitfile "${junit_xml_filename}" --raw-command cat "${junit_filename_prefix}"*.stdout
-  if [[ ! ${KUBE_KEEP_VERBOSE_TEST_OUTPUT} =~ ^[yY]$ ]]; then
-    rm "${junit_filename_prefix}"*.stdout
-  fi
 
   if ! command -v prune-junit-xml >/dev/null 2>&1; then
     kube::log::status "prune-junit-xml not found; installing from ./cmd"
     go -C "${KUBE_ROOT}/cmd/prune-junit-xml" install .
   fi
-  prune-junit-xml "${junit_xml_filename}"
-
-  kube::log::status "Saved JUnit XML test report to ${junit_xml_filename}"
 }
 
 runTests() {
   local junit_filename_prefix
   junit_filename_prefix=$(junitFilenamePrefix)
 
-  # Try to normalize input names.
+  installTools
+
+  # Try to normalize input names. This is slow!
   local -a targets
+  kube::log::status "Normalizing Go targets"
   kube::util::read-array targets < <(kube::golang::normalize_go_targets "$@")
 
   # If we're not collecting coverage, run all requested tests with one 'go test'
   # command, which is much faster.
   if [[ ! ${KUBE_COVER} =~ ^[yY]$ ]]; then
     kube::log::status "Running tests without code coverage ${KUBE_RACE:+"and with ${KUBE_RACE}"}"
-    # shellcheck disable=SC2031
-    go test "${goflags[@]:+${goflags[@]}}" \
-     "${KUBE_TIMEOUT}" "${targets[@]}" \
-     "${testargs[@]:+${testargs[@]}}" \
-     | tee ${junit_filename_prefix:+"${junit_filename_prefix}.stdout"} \
-     | grep --binary-files=text "${go_test_grep_pattern}" && rc=$? || rc=$?
-    produceJUnitXMLReport "${junit_filename_prefix}"
+    gotestsum --format="${gotestsum_format}" \
+              --jsonfile="$(if [[ -n "${junit_filename_prefix}" ]] && [[ ${KUBE_KEEP_VERBOSE_TEST_OUTPUT} =~ ^[yY]$ ]]; then echo "${junit_filename_prefix}.stdout"; fi)" \
+              --junitfile="${junit_filename_prefix:+"${junit_filename_prefix}.xml"}" \
+              --packages="${targets[*]}" \
+              -- \
+              "${goflags[@]:+${goflags[@]}}" \
+              "${KUBE_TIMEOUT}" \
+              -args "${testargs[@]:+${testargs[@]}}" \
```

Comment:
Somehow `-short` gets added to `testargs` during "make test-integration", breaking pull-kubernetes-integration. It needs to be in `goflags`. Will fix...

Reply:
hmm... it gets put in KUBE_TEST_ARGS, which should only be used as args to `go test`, where `-short` is valid, right?

Reply:
oh, *this* PR is changing the `go test` invocation to `gotestsum` ... seems like `gotestsum` must support the same surface area as `go test` to do that

Reply:
yeah, this sort of thing is why we don't do this already, so we don't lost control of the actual test invocation.

Reply:
(ref: https://github.com/kubernetes/kubernetes/pull/80822)

Reply:
> it gets put in KUBE_TEST_ARGS, which should only be used as args to go test, where -short is valid, right?

It's documented as "KUBE_TEST_ARGS: Arguments to pass to the resulting go test invocation."

https://github.com/kubernetes/kubernetes/pull/107516 rearranged it so that `-args <test binary arguments>` can be added at the end.

That works, but only when being very careful about where to put those arguments. I wish there was a clearer separation between arguments for `go test` and arguments for the test binary, but it is what it is.

This means that with `gotestsum` we need to use `--raw-command` and put together the final `go test` invocation the same way as it was done before.

Reply:
It's working now:
- https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/125534/pull-kubernetes-unit/1818911988575637504
- https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/125534/pull-kubernetes-integration/1818911985216000000

The amount of unnecessary log output now stands out more than it did before. Something to tackle next...